src/ngx_http_lua_socket_tcp.c

```diff
@@ -1943,6 +1948,38 @@ ngx_http_lua_socket_read_all(void *data, ssize_t bytes)
 }
 
 
+static ngx_int_t
+ngx_http_lua_socket_read_bsd(void *data, ssize_t bytes)
+{
+    ngx_http_lua_socket_tcp_upstream_t      *u = data;
+
+    ngx_buf_t                   *b;
+
+#if (NGX_DEBUG)
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, u->request->connection->log, 0,
+                   "lua tcp socket read bsd");
+#endif
+
+    if (bytes == 0) {
+        u->ft_type |= NGX_HTTP_LUA_SOCKET_FT_CLOSED;
+        return NGX_ERROR;
+    }
+
+    b = &u->buffer;
+
+    dd("already read: %p: %.*s", u->buf_in,
+       (int) (u->buf_in->buf->last - u->buf_in->buf->pos),
+       u->buf_in->buf->pos);
+
+    dd("data read: %.*s", (int) bytes, b->pos);
+
+    u->buf_in->buf->last = ngx_cpymem(u->buf_in->buf->last, b->pos, bytes);
```

Comment:
`u->buf_in->buf` and `b` actually point to the same buffer. So there is no need to copy the content. Just increase the pointer position is enough.